### PR TITLE
Assignment 2

### PR DESCRIPTION
**Pull request by lthurair**
## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)
added code to labs 4-6 and assignment 2
## What did you learn from the changes you have made?
learned to create neural network
## Was there another approach you were thinking about making? If so, what approach(es) were you thinking of?
n/a
## Were there any challenges? If so, what issue(s) did you face? How did you overcome it?
n/a
## Checklist
- [X] I can confirm that my changes are working as intended

### CODE REVIEW

File: 01_materials/labs/lab_4.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_4.ipynb:1]: - A simple convolution filter: You defined a `Conv2D` with kernel size 7 but initialized it with kernel size 5 (my_kernel). Also, the reason we used `1/25` in the example is the sum of the filter is `25` and we can *normalize* the filter with its sum to avoid overflow (>255). Using `1/125` isn't wrong but as you can see the image becomes darker because the values are getting smaller when divided by 125.
- Using a pretrained model: I personally suggest pre-processing the image with `preprocess_input()` without `resize()` when using a pretrained model, since it may introduce artifacts to the image and lead to strange predictions. `preprocessing_input()` will resize the image automatically.

File: 01_materials/labs/lab_5.ipynb
Comment (tianyi21) [inline 01_materials/labs/lab_5.ipynb:1]: - Build a better model: You may remove `if class_prediction_head and box_prediction_head:` to train the better model. Also, you included `relu` as activation function in the `box_prediction_head` whereas you gave an explanation on why we shouldn't use activations.

File: 02_activities/assignments/assignment_2.ipynb
Comment (tianyi21) [inline 02_activities/assignments/assignment_2.ipynb:1]: - Task 3.3: Instead of finding the most likely word, you may consider defining a probability distribution with the predicted class probability, which will help you break loops in text generation.
```
# Most likely
predicted_word_index = np.argmax(yhat) 
# Probability distribution
yhat = np.random.choice(range(VOCAB_SIZE), p=yhat.ravel())
```
- Model Refinement: Please also try to generate the text with your improved model.

**Review by tianyi21 — APPROVED** [review]
Hi Lisa!

Thank you for submitting your DL assignments. Please refer to my comments attached to each file.

All your DL assignments have been completed. Thank you for your participation, and I wish you all the best for your future!

Thanks,
Tianyi [LS]

**Comment by lthurair** [comment]
Thank you Tianyi! I will make those changes. Best wishes to you too!
